go/host/db/batches.go

```
@@ -0,0 +1,112 @@
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+)
+
+// DB methods relating to batches.
+
+// headerKey = batchHeaderPrefix  + hash
+func batchHeaderKey(hash gethcommon.Hash) []byte {
+	return append(batchHeaderPrefix, hash.Bytes()...)
+}
+
+// GetHeadBatchHeader returns the header of the node's current head batch, or (nil, false) if no such header is found.
+func (db *DB) GetHeadBatchHeader() (*common.Header, bool) {
+	headBatchHash, found := db.readHeadBatchHash(db.kvStore)
+	if !found {
+		return nil, false
+	}
+	return db.readBatchHeader(db.kvStore, *headBatchHash)
+}
+
+// AddBatchHeader adds a batch's header to the known headers
+func (db *DB) AddBatchHeader(header *common.Header, txHashes []common.TxHash) {
+	b := db.kvStore.NewBatch()
+	db.writeBatchHeader(b, header)
+
+	// TODO - #718 - Store the batch txs, batch hash, and batch number per transaction hash, if needed (see `AddRollupHeader`).
+
+	// There's a potential race here, but absolute accuracy of the number of transactions is not required.
+	currentTotal := db.readTotalTransactions(db.kvStore)
+	newTotal := big.NewInt(0).Add(currentTotal, big.NewInt(int64(len(txHashes))))
+	db.writeTotalTransactions(b, newTotal)
+
+	// update the head if the new height is greater than the existing one
+	headBatchHeader, found := db.GetHeadBatchHeader()
+	if !found || headBatchHeader.Number.Int64() <= header.Number.Int64() {
+		db.writeHeadBatchHash(b, header.Hash())
+	}
+
+	if err := b.Write(); err != nil {
+		db.logger.Crit("Could not write batch.", log.ErrKey, err)
+	}
+}
+
+// Retrieves the batch header corresponding to the hash, or (nil, false) if no such header is found.
+func (db *DB) readBatchHeader(r ethdb.KeyValueReader, hash gethcommon.Hash) (*common.Header, bool) {
+	f, err := r.Has(batchHeaderKey(hash))
+	if err != nil {
+		db.logger.Crit("could not retrieve batch header.", log.ErrKey, err)
```

Error should be returned

---

Not swallowed - Crit() causes the program to exit.

---

Sorry, I mean that the error isn't being returned, swallowed was the wrong word, but it was in the sense that the caller doesn't have to handle it.

---

Yeah, the approach we’ve chosen here is for the process to exit if the db is unreachable or has a critical issue. Maybe I can add a todo to handle critical db errors gracefully?

---

Raised https://github.com/obscuronet/obscuro-internal/issues/1195. Will tackle after.